Add win32-aarch64 support

tresf opened the pull request:
#### TODO
* [x] Review by jna team
* [x] Fix compiler warnings
* [x] Review unit tests
* [x] Discuss libffi changes/upstream contributions (Relevant discussion: https://github.com/java-native-access/jna/issues/1259#issuecomment-721856499)
* [x] Possibly add [AppVeyor](https://github.com/java-native-access/jna/blob/0a0e44d4e8c616eb948fae7b87f882db32db5e60/appveyor.yml#L12) support?
* [x] Documentation
* [x] Fix x86 builds https://github.com/java-native-access/jna/pull/1264#issuecomment-725626587 (possibly https://github.com/libffi/libffi/issues/71?)
* [x] Fix newly failing +173 tests (e.g. [`testSyncronizedAccess` unit test](https://github.com/java-native-access/jna/pull/1264/commits#issuecomment-735029344)).  (Upstream regression: https://github.com/libffi/libffi/pull/593#issuecomment-735451629.  Reverted.)
* [x] Fix unexpected line-ending in `native/libffi/libtool-ldflags` (https://github.com/java-native-access/jna/pull/1264/commits/97a66a78264602b60b57cb4f7aa5b6c8d18bd921)


#### Changes
* Updates `native/libffi` to [latest version, adding Windows arm64 support](https://github.com/libffi/libffi/pull/486) 2bf1593
* Disables DLL callback functionality for `win43-aarch` via [`ASMFN_OFF`](https://github.com/java-native-access/jna/blob/2d9630516081508979055e155f5dfd34c9fc13d0/native/Makefile#L471) (no mingw support)
* Disables DLL callback tests for `win32-aarch64` 2ff29eb
* Fixes marshal test for `win32-aarch64` e3ad29d
* Patches `msvcc.sh` to handle the following scenarios/regressions:
   * Re-adds mvvcc.sh's `cygpath` calculation for `cl.exe` 112d76d
   * Quotes mvvcc.sh's include (`-I`) files with spaces (e.g. `-I"C:\Program Files\AdoptOpenJDK11\...\include`) 112d76d
   * Brings back mvvcc.sh's `eval` for assembly files 112d76d
* Adds `win32-aarch64` native build option
   * Toggled via:
      ```bash
       ant native -Dos.prefix=win32-aarch64
      ```
   * Sets `${make.ARCH}=aarch64`, reusing Android's property
   * Produces new artifact `lib/native/win32-aarch64.jar`
   * Tested build using `MSVC v142 - VS 2019 C++ ARM64 build tools (v14.27)` on Windows 10 x86_64
   * Tested binary using Surface Pro X and [`openjdk-aarch64` from Microsoft](https://github.com/microsoft/openjdk-aarch64/)

#### Steps

See the updated [`www/WindowsDevelopmentEnvironment.md`](https://github.com/java-native-access/jna/blob/5fbebe0e2034ec745aa1cedf82a4b4f99f2c4411/www/WindowsDevelopmentEnvironment.md)

---

<details>
<summary>Hidden, outdated</summary>

* [x] ~~Bump `jna.jar`~~ No, per https://github.com/java-native-access/jna/pull/1264#issuecomment-731357977

---

   1. Install and configure `ant` for your system
   1. Install cygwin per [Windows steps](https://github.com/java-native-access/jna/blob/master/www/WindowsDevelopmentEnvironment.md#mingw)
   1. Install Visual Studio 2019 with ARM build tools (other versions should also work).
   1. Disable CRLF globally for git (use with caution, this will affect all projects.  Alternately, you can skip this step and use `dos2unix` on each failing file)
      ```bash
      git config --global core.autocrlf false
      git config --global core.eol lf

      # Fixes error: AC_CONFIG_MACRO_DIRS([m4]) conflicts with ACLOCAL_AMFLAGS=-I m4
      # See also: https://stackoverflow.com/q/47582762
      ```
   1. Clone this PR
      ```bash
      git clone -b win-arm64 https://github.com/tresf/jna --depth 10
      cd jna
      ```

   1. From a Command Prompt, configure MSVC to build using arm64
      ```bash
      "%ProgramFiles(x86)%\Microsoft Visual Studio\2019\Community\Common7\Tools\VsDevCmd" -arch=arm64 -host_arch=amd64

      # or on native aarch64 use "-host_arch=x86"
      ```

   1. Append cygwin to your path (Append, **DO NOT** prepend, if you prepend, `link.exe` will fail!)
      ```bash
      set PATH=%PATH%;C:\cygwin64\bin\

      # or on native aarch64, "C:\cygwin\bin"
      ```

   1. Fire the build:
      :warning: **Note:** Windows Defender real-time protection makes this very, very slow.  Consider disabling or adding an exception to the `jna` directory.
      ```bash
      ant native -Dos.prefix=win32-aarch64

      # or on native aarch64, just type "ant"
      ```

   1. Clean between errors:
      ```cmd
      ant clean
      ```

   1. Optionally, enable debugging
      * Change `verbose=` to `verbose=1` in `native/libffi/msvcc.sh`
      * Append `-DEXTRA_MAKE_OPTS="--debug=v"` to ant command

</details>

tresf commented:
Some CI failures seem to be related to what was described as ["flaky Travis behavior"](https://groups.google.com/d/embed/msg/jna-users/15_k9BtvUOQ/xQpxSnyMBgAJ) on the mailing list, specifically `X11Test.java:26: error: cannot find symbol com.sun.jna.StructureFieldOrderInspector;`.  Another appears to be a missing tool, `textinfo`, which I'll attempt to add as an additional CI dependency.

matthiasblaesing commented:
I did not yet really check, but I managed to get the build for aarch64 going. I have now three different dlls for windows, for two of these the `file` tool can tell me the architecture, for one not. This is to be expected, as aarch64 is very new and I assume file was not yet updated.

Unittests for x86-64 come back clean, but unittests for x86 result in an error:

```
    [junit] Testcase: testStdCallCallbackStackAlignment(com.sun.jna.win32.W32StdCallTest):      FAILED
    [junit] stdcall callback did not restore the stack pointer
    [junit] junit.framework.AssertionFailedError: stdcall callback did not restore the stack pointer
    [junit]     at com.sun.jna.win32.W32StdCallTest.testStdCallCallbackStackAlignment(W32StdCallTest.java:180)
    [junit]
```

In the dicussion for #1259, some patches were mentioned, that match this: https://github.com/java-native-access/jna/issues/1259#issuecomment-720634705.

You can run the build + unittests by invoking:

```
ant
```

On x86-64 you can switch between 32bit and 64bit by switching the building JDK. When run on aarch64 this should also be possible to run through the build and check the unittests.

Thank you for the work. It is great, that you took this on.

tresf commented:
AppVeyor steps for x86, arm64 added via 3061cd3.  x86 is failing at the same spot as identified in https://github.com/java-native-access/jna/pull/1264#issuecomment-725626587.

matthiasblaesing commented:
For the texinfo problem on travis this should fix it: https://github.com/matthiasblaesing/jna/commit/b69871043030a9ee125b9b9062c7f6e5f705f50c (test on travis-ci on amd64 are currently dog-slow, so this is deduced from the successful run on arm64)

tresf commented:
@matthiasblaesing thanks, I'll rebase once merged.

tresf commented:
> You can run the build + unittests by invoking:
> ```bash
> ant
> ```

Done.  Please see test results:

![image](https://user-images.githubusercontent.com/6345473/99128065-d368f700-25d7-11eb-881c-5724320fc960.png)

Detailed logs can be found here: https://gist.githubusercontent.com/tresf/dab4ebf5459f3370a33c0c0c659e5d4e/raw/788512b93e2a58eb7cc18aa171d5da972fdcdbec/win32-aarch64-jna-tests.log

@matthiasblaesing thoughts?

Some updates to the original build instructions:

* No changes to the build system were needed.  With the an aarch64 build of Java `ant` works just fine.
* Instructions have been updated with instructions for building on Windows for ARM64 (mostly just switching the toolchain over to x86 and removing the `os.prefix`).
* Building on native ARM64 hardware is quite slow (2x - 3x slower than Intel) as it uses MSVC compiler through x86 emulation to cross-compile aarch64 binaries.  Added note about disabling Windows Defender to speed this up a bit.

matthiasblaesing commented:
For the arguments marshal tests, this should be the relevant commit, that fixed it in the past:

- Original Fix: https://github.com/java-native-access/jna/pull/1122
- Adaption for ARM64: https://github.com/java-native-access/jna/commit/5c889be

For the test failure in "GetModuleHandleEx(fptr) failed":

If I remember correctly you disabled that code for arm64 didn't you? If so it needs to be recorded in `Platform`: `com.sun.jna.Platform.HAS_DLL_CALLBACKS`. Currently windows is mapped to `true`. This needs an additional constraint:

```java
        HAS_DLL_CALLBACKS = osType == WINDOWS;
```

For the crash in `LoadLibraryTest` a look into: C:\Users\Owner\jna\hs_err_pid18788.log would be interesting.

The error in testLoadFromUnicodePath is more or less expected - it works only on the "right" OS encoding.

tresf commented:
> The error in `testLoadFromUnicodePath` is more or less expected - it works only on the "right" OS encoding.

I tried `set _JAVA_OPTIONS=-Dfile.encoding=UTF-8` and the unit tests pick it up, but they still fail.  I'm sure this is well-known so I won't waste much time on it unless needed by the project.

> marshal tests [...] adaption for ARM64: 5c889be

Perfect, thanks.  Patched e3ad29d.

> `com.sun.jna.Platform.HAS_DLL_CALLBACKS` needs an additional constraint:

Thanks, patched via 2ff29eb.

> For the crash in `LoadLibraryTest` a look into: [crash log] would be interesting.

Here's the crash log: https://gist.githubusercontent.com/tresf/caca6d16bfde240c87d17d72194564c2/raw/43d9c8552dc627894b877206e11229ad68bec3f8/hs_err_pid19056.log

If I need to rebuild with debug symbols, I'd be happy to.  I'm not very familiar with this process, but I'd be happy to try.

New log here: https://gist.githubusercontent.com/tresf/0f183d54d504134c953a190db46b8ffb/raw/ae7cac174641339017f1058842d25666cbcd8497/win32-aarch64-jna-tests2.log


New report:

![image](https://user-images.githubusercontent.com/6345473/99196363-e7d4fd00-2759-11eb-8e91-2726c4e33ebd.png)

tresf commented:
Update.  The unit tests are in better shape after updating to the [latest win32-aarch64 snapshot from Microsoft](https://github.com/microsoft/openjdk-aarch64/releases), now called '16-ea`, which specifically addresses unrecoverable crashes, quoting:

> Contains fix for VEH exception handling, so that HotSpot will resume exception handling instead of bailing out, so that a potential SEH handle of a native library has a chance to catch it. Instead, we install an unhandled exception handle.

... more release notes are available in the [releases area](https://github.com/microsoft/openjdk-aarch64/release).  Note, at the time of writing this the newest release is for macOS (notice the `.tar.gz` extension), so please ignore that build specifically. 

New results attached.

Logs: https://gist.github.com/tresf/07c4de8cbbf17d87997da261dae40d7b

Test summary:
![image](https://user-images.githubusercontent.com/6345473/99197128-85323000-275e-11eb-92c7-ded7c8192bfe.png)

tresf commented:
Note, the following upstream issues/PRs seem to be very similar to the x86 issues we see in the `testStdCallCallbackStackAlignment`.  The first (and oldest is JNA's own from @twall).

Listed as a code block to avoid a bunch of crosslinking.
```
https://github.com/libffi/libffi/issues/198 - win32-x86 structure handling broken 
https://github.com/libffi/libffi/issues/215 - win32 x86 stdcall closure: incorrectly restored stack after closure call 
https://github.com/libffi/libffi/pull/514 - Clang on windows uses 4 byte stack alignment, not 16
https://github.com/libffi/libffi/pull/378 - Fix FFI_STDCALL ABI
https://github.com/libffi/libffi/pull/465 - Fix Win32 stdcall closure
```

I cherry-picked what I think is the same x86 guard.  The diff is below.  With this applied, `testStdCallCallbackStackAlignment` passes.  I'm not sure if MSVC check is warranted too.  This is out of my area of expertise.


```diff
diff --git a/lib/native/win32-x86.jar b/lib/native/win32-x86.jar
index 54fab71e8..025ca3e89 100644
Binary files a/lib/native/win32-x86.jar and b/lib/native/win32-x86.jar differ
diff --git a/native/libffi/src/x86/ffi.c b/native/libffi/src/x86/ffi.c
index 5f7fd81d9..98c09ba6a 100644
--- a/native/libffi/src/x86/ffi.c
+++ b/native/libffi/src/x86/ffi.c
@@ -181,6 +181,9 @@ ffi_prep_cif_machdep(ffi_cif *cif)
     {
       ffi_type *t = cif->arg_types[i];

+#if defined(_M_IX86)
+      if (cif->abi != FFI_STDCALL)
+#endif
       bytes = FFI_ALIGN (bytes, t->alignment);
       bytes += FFI_ALIGN (t->size, FFI_SIZEOF_ARG);
     }
```

tresf commented:
The above patch seems to work with both mingw and msvc builds, I'm going to add it to this PR.  Next, I'll rebase and cleanup the commits slightly.

tresf commented:
@matthiasblaesing some questions:

* Would you like me to bump `jna.jar` with this PR, or would you rather a project maintainer do that?
* Would you like me to add any documentation to `www/WindowsDevelopmentEnvironment.md`?
* I'm not well versed in how the callback code is handled in 004c975.  Would you or @twall or another project maintainer be willing to help with the upstream patch?  I can propose the change, but I don't feel comfortable enough with defending it as I do not fully understand the code..  Reading related commits mentioning same target using different compilers makes me even more uncomfortable (e.g. not limiting the patch to MSVC, etc).
* Last, since `libffi/libffi` has -- and likely will continue to -- suffer from divergence issues, are there any further requests for handling that or would you prefer to tackle that another time?

I'm happy to submit an upstream patch for the `msvcc.sh` at any time.

matthiasblaesing commented:
> Would you like me to bump `jna.jar` with this PR, or would you rather a project maintainer do that?

I assume, that you mean `dist/jna.jar`. Then no, that file is updated when a release is done. The release process builds the java part from source and then bundles the binaries for the platforms, forming the final `jna.jar`, which is then also published to maven central.

> Would you like me to add any documentation to `www/WindowsDevelopmentEnvironment.md`?

I use that file as my base to rebuild the native binaries. If I understood correctly, the build works in cross-platform mode and thus it would be great to add a description. At some point someone needs to rebuild the native parts and then these notes. It is best if the binary is rebuild on the target platform itself, but my experience is, that this is not always possible.

> I'm not well versed in how the callback code is handled in [004c975](https://github.com/java-native-access/jna/commit/004c97533bb99a49ae83c4629796f90dace28121).  Would you or @twall or another project maintainer be willing to help with the upstream patch?  I can propose the change, but I don't feel comfortable enough with defending it as I do not fully understand the code..  Reading related commits mentioning same target using different compilers makes me even more uncomfortable.

Sorry, I'm also more at home on the java side. I can fix obvious problems and spot problems, but here I'm out. It would indeed be great if you could help with upstreaming @twall. In general the current state looks sane to me and I'm inclined to move take it. I did a diff against upstream master and there is some movement in the ARM64/windows area, so it might be worth updating while its hot :-).

> Last, since `libffi/libffi` has -- and likely will continue to -- suffer from divergence issues, are there any further requests for handling that or would you prefer to tackle that another time?

If people want to do it in the JNA repository, ok, but then the question needs to be answered how sustainable that is. If we have changes separated out, documented and covered with tests, that are run as part of the build, all is well. If not I think it is better to push changes to go through upstream.

I like the debian approach, where upstream is augemented with seperated patches, which can be reviewed/exchanged individually.

<hr />

I tested this branch on windows-x86-64, windows-x86 and linux-x86-64. Build runs clean and unittests run clean (that means the unittests that fail on master also fail on this PR, no new failures are reported). 

I have a few nitpicks for the commits:

bd672b9f137970859cb1ddadef91a8acdcb4df80: It would be great if the version/commit of libffi could be mentioned that was used to create that update. That might be helpful if future developers try to do an update.

cef2d768efe533b56f5ef33c1321abc3d1348182 + 9f0775bd0658ba5a81baf9e718bdab8c1fe7ea7a: Please squash. It should not matter, as the win32-aarch64.jar only contains native code, so should be save anyway.

Please update your author information for the commits, to contain your fullname. The email part is already spells it out, but I'd like to keep the author information correct.

Please add an entry to CHANGES.md about the update of libffi and more importantly the added platform support.

tresf commented:
> It is best if the binary is rebuild on the target platform itself, but my experience is, that this is not always possible.

Since Microsoft doesn't offer a native ARM64 version of Visual Studio, the target platform uses near-identical build tools.  The major difference is the `x86` versus `x86_64` toolchain.

Since..
1. Microsoft [will soon be adding `x86_64` emulation support to the ARM64 version of Windows](https://blogs.windows.com/windowsexperience/2020/09/30/now-more-essential-than-ever-the-role-of-the-windows-pc-has-changed/)
1. Visual Studio 2019 finally defaults to a 64-bit toolchain.
   * Note, from what I'm reading, [Visual Studio 2017 also supports ARM64](https://pete.akeo.ie/2017/05/compiling-desktop-arm-applications-with.html)
1. Consistent with changes to AppVeyor
1. The compiler and toolchain would then be identical between native and cross.
... I'm inclined to keep the tutorial for 64-bit systems.  Do you agree?  If not, we get into some differentiating steps between `cygwin` versus `cygwin64`,   I can spell these out if needed.

That said, the binary provided was built on an Intel machine.  That's purely for convenience and **I'm happy to re-upload using a build created with the 32-bit toolchain on native hardware**.  I just wanted to clarify that the native build still uses an x86 environment until Microsoft releases a version of Visual Studio for ARM64.

> Please update your author information for the commits, to contain your fullname.

Happy to, but where do you see this missing information?

In regards to libffi/upstream would jna consider having their own fork?  I can help setup the submodule, but developers would need to do `git clone ... --recursive` each time, which might be frustrating, but at least there'd be a clear, historical divergence path and the PRs would likely originate from JNA proper, instead of specific developer branches.  I've done this [quite a bit](https://tinyurl.com/y6oksb4r) and happy to help with this effort, if interested.

Regardless of that decision, in the interim, I'll reword the commit to include the specific versioning information and address the other requests as well.

tresf commented:
Initial rewrite of the readme is here: https://gist.github.com/tresf/9182a81d7c41a9425ae8ad8f626773e1

Please let me know if this is in the right direction, or if you'd prefer to keep the `<pre>` style with one section per-arch.  I've deliberately hidden the old instructions as your notes were far easier to follow.

tresf commented:
I believe all request have been addressed with exception of an updated Windows procedure.  When I have direction on that, I'm happy to commit or squash.

matthiasblaesing commented:
Thank you for the updates. The new build instructions look sane. I'll try to test them in the next few days. I plan to reset my windows development setup. Two things I noticed:

1. You describe the build of the x86 binaries with mingw64. Were you successful with that (i.e. did run all the unittests for JNA and platform cleanly?) I remember that I had issues with x86 binaries build with cygwin64, but I could be mistaken and will retry.
2. For the crossbuild I think you will need an openjdk for windows aarch64, else you might pick up definitions, that don't match aarch64, but amd64

tresf commented:
> The new build instructions look sane.

Thanks, pushed so mistakes or changes can be commented on directly in the PR.

> You describe the build of the x86 binaries with mingw64. Were you successful with that (i.e. did run all the unittests for JNA and platform cleanly?) I remember that I had issues with x86 binaries build with cygwin64, but I could be mistaken and will retry.

Yes, 8eac2a5 was performed using cygwin64.  As long as the JVM is 32-bit the unit tests seem OK (minus the encoding error, which I'd love to tackle too, but couldn't figure out).

> For the crossbuild I think you will need an openjdk for windows aarch64, else you might pick up definitions, that don't match aarch64, but amd64

Correct and this is vaguely hinted [here](https://github.com/java-native-access/jna/pull/1264/commits/5fbebe0e2034ec745aa1cedf82a4b4f99f2c4411#diff-09e6c361637588838c5e94735de1c6d78a0c7879165237126fd9ced7f8659da6R74).  The founders of [AdoptOpenJDK were hired by Microsoft](https://jaxenter.com/adoptopenjdk-contributor-jclarity-acquired-microsoft-161112.html) but the `aarch64` backports to JDK8, JDK11 are [quite a ways out](https://tinyurl.com/y3r3cgvs) so I'm not sure how soon they'll land in AdoptOpenJDK's website, and there's no dedicated download page, so I was intentionally vague.  Happy to improve as needed.

tresf commented:
**Edit:** ... and for those with Intel hardware, using the `os.prefix=win32-aarch64`, the build fires fine on Intel, but no unit tests capabilities. The x86_64 JVM seemed fine with this although I see now that `os.prefix=win32-aarch64` isn't documented and I'd happily add this flag to the documentation if warranted.

fkistner commented on build.xml:
```diff
@@ -876,7 +882,7 @@ osname=macosx;processor=x86;processor=x86-64;processor=ppc
     </condition>
     <condition property="ARCH" value="aarch64">
-      <equals arg1="${os.prefix}" arg2="android-aarch64"/>
+      <matches string="${os.prefix}" pattern="-aarch64$"/>
```
Note that this is also duplicated in `native/build.xml`.

tresf replied:
@fkistner, we couldn't find any documentation or evidence that `native/build.xml` was even used by the project, but I'll patch that up now for consistency.

tresf replied:
Done. https://github.com/java-native-access/jna/pull/1264/commits/a5e50c6ec1d721ef778e8d6f913cb369c900da9b#diff-28d14c8a63382329901de86493b3db46d5447a0a17f37f240b65687dc203b749R239

tresf commented:
Upstream has merged the modifications to `msvcc.sh` via https://github.com/libffi/libffi/pull/596.

matthiasblaesing commented:
Sorry for the quietness. This is what happened in the mean time: I rebuild locally the windows x86 + x86-64, linux-64, solaris (sun+x86), aix (32+64bit). On aix I had to update the gnu build tools.

Before the release I'll rebuild all native libraries, but as mac os aarch64 is also in flight, I'd like to get that in also.

I checked the includes for x86-64 and aarch64 and the JNI defines are identical (there are differences in jvmti, but that should not affect us), so crossbuilding on x86-64 for aarch64 should be safe.

I checked the libffi update and noticed, that the reference is not correct. The merged state of libffi is 8111cd06921e80d5d7192ce8d1f64733072fdbcd. 

I suggest to either update the commit message to record the correct commit or bring the embedded libffi to current master, this might also be helpful for the macos aarch64 update.

For the build instructions, I would remove the directions how to build on 32 bit windows and instead stress, that building should happen on 64bit.

With that done I'd merge, we should focus on mac os aarch64, stabilize the tree, rebuild the natives and try to get a release out.

tresf commented:
> I would remove the directions how to build on 32 bit windows

Done, just a word of caution, until x86_64 emulation is offered on ARM64 Windows, builds on that hardware will require these additional, undocumented steps.  They've been removed from the Troubleshooting area.

> instead stress, that building should happen on 64bit.

Added "64-bit" to "Starting point:".  This statement should be true for both aarch64 and x86_64 hosts.

>  libffi update and noticed, that the reference is not correct

Rebased jna, bumped libffi and corrected.  Also bumped `lib/native/win32-aarch64.jar`.  Bumping removes a commit -- our downstream changes to `msvcc.sh.  Unfortunately it causes a newly failed unit test:

```diff
-    [junit] Testcase: testSynchronizedAccess(com.sun.jna.NativeTest):   FAILED
-    [junit] NativeLibrary lock should be held during native call to synchronized library
-    [junit] junit.framework.AssertionFailedError: NativeLibrary lock should be held during native call to synchronized library
-    [junit]     at com.sun.jna.NativeTest.testSynchronizedAccess(NativeTest.java:219)
-    [junit]     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
-    [junit]     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
-    [junit]     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
```

matthiasblaesing commented:
@tresf thank you. I ran tests on linux-x64, windows-x64 and windows-x86. There errors reported there and on appveyor are expected (on my system) or explainable (on appveyor 32 bit builds). I could not reproduce the problem.

Please double check - if the problem is reproducible, I'd merge if you agree.

tresf commented:
It turns out the errors are +173, not +1 as I had originally reported.  I'm not sure the cause, but I believe something's gone awry, perhaps with the rebase or `dos2unix|newline branch` removal.  I'll spend some time figuring out exactly the cause.

tresf commented:
I believe I corrupted my own `libffi` directory.  I disabled `eof=lf` globally, copied `libffi/libffi` into `native/libffi` and this resulted in some corrupted files.  For example, in my local, I have the following error:

```diff
- /cygdrive/c/Users/Owner/jna/native/libffi/libtool-ldflags: line 2: $'\r': command not found
```

This is despite `.gitattributes` treating this as `eol-lf` per #1273.

This problem won't go away, any time we bump `native/libffi` from a Windows machine, we'll be pestered with more of these line-ending problems.  A submodule may fix this, assuming upstream has a `.gitattributes` that's comparable to ours.

For now, I'll attempt the bump again from a Unix machine and see if the issue go away.

tresf commented:
Ok, I believe I know what's happening... 

`libtool-ldflags` is explicitly listed to use `eol=lf`, matched on [`#1273#Ll9`](https://github.com/java-native-access/jna/pull/1273/files#diff-618cd5b83d62060ba3d027e314a21ceaf75d36067ff820db126642944145393eR9), but it's ignored.  The repo shows it using `lf`, so I assume this is because [`.gitignore` mentions this file by exact name](https://github.com/java-native-access/jna/blob/master/native/libffi/.gitignore#L23).  I assume this is an edge-case git bug.

There are a few ways to fix this:
1. (Preferred) Remove `libtool-ldflags`, hope it gets reconfigured when native compilation is fired<br>--OR--
2. Comment-out `libtool-ldflags` in `.gitignore`, causing even more divergence issues<br>--OR--
3. Force entire project to just always use `eol=lf`, might break `.sln` files

Running some tests now to see which is most viable.

tresf commented:
> If this file is not in the repository, it is not copied, so git handling has nothing to do with this.

It's a file that `libffi/libffi` upstream explicitly has in the `.gitignore` but ships anyway.  It appears `git` ignores the `.gitattributes` for files listed in the `.gitignore`.  I assume it's an actual git bug, but it's so edge-case, I can't find any mentions of it.  Commenting this out in `95750d2` unintuitively fixes the `\r` errors which `libtool-ldflags` creates.

Unfortunately, this does nothing to remedy the failed unit-tests, so I'll do more digging.

tresf commented:
I believe this was broken by https://github.com/libffi/libffi/pull/593.  Reverting that commit seems to bring the unit-tests back into workable order.

I also believe `libtool-ldflags` can safely be left out.

Since our upstream changes were merged after https://github.com/libffi/libffi/pull/593, I'll have to revert back to:

> The merged state of libffi is 8111cd06921e80d5d7192ce8d1f64733072fdbcd.

Starting that now to make this library usable.  Once working, I'll ping the author of https://github.com/libffi/libffi/pull/593 to see if they have any insight as to why their PR causes regressions on Windows aarch64.

matthiasblaesing commented:
Just to get me up to speed - you are seeing test failures only on aarch64?

tresf commented:
> Just to get me up to speed - you are seeing test failures only on aarch64?

Yes.

Although I'm only testing on `aarch64`.  The plan was to bring upstream libffi changes in (get rid of our own https://github.com/java-native-access/jna/commit/112d76d4e2feb2b21c36f6753389bb2b88f8acb2), but that lands upstream after libffi/libffi#593 which I believe to be causing this serious regression.

Now I will fixup this PR so that jna is back to mergeable state and not waiting on upstream.  In the meantime if upstream is quick to fix, I can perform the fast-forward again.

tresf commented:
It appears this project has always had `libtool-ldflags` in the source, so I believe removing it was a mistake.  This means we'd otherwise have to  patch it after cloning if a Windows system is configured with default line-endings.

For that reason, I'm going back to commenting it out in the `.gitignore`.  Ugh.  Sorry, more force-pushes incoming.

tresf commented:
Please pardon the force-pushes.  `libffi` pollutes it's own build directory so adding untracked files and re-testing a clean clone is a very time consuming process.

tresf commented:
@matthiasblaesing I believe this is ready for final review.  Cloning on Windows without `core.eol lf` succeeds with the addition of https://github.com/java-native-access/jna/commit/97a66a78264602b60b57cb4f7aa5b6c8d18bd921 and I've proposed the change upstream in hopes it'll become default.

tresf commented:
After studying the source code more, I've come to the conclusion that the callback functions can be avoided without building the stub library, simplifying the codebase.

I'm going to send some patches in to simplify this.  I think they'll be well received.  They'll be in-addition-to for proper review, then I can squash them if there are no objections.

tresf commented on native/Makefile:
```diff
@@ -246,8 +246,21 @@ LIBS=psapi.lib
+FFI_CONFIG+=--host=aarch64-cygwin
+# Skip building assembly callback functions
+CDEFINES+= -DASMFN_OFF
+DLLCB=
```
So the new fix is a bit more elegant.  If we wipe out `DLLCB` var, the `EXTRAOBJS+=` lower in the Makefile won't add it, which is how non-`amd64` does it anyway.  The custom precompiler flag is still needed.  We could detect an aarch64 compiler, but it's more controlled this way and allows a farily quick toggle-off when mingw adds aarch64 support.

matthiasblaesing commented:
I'd like to see the problem with libffi master resolved before this is merged. It might be helpful to see a full build log from a failing build with libffi master. Could it be that aarch64 on non-linux platforms is still in high flux? The indication, that MS decided not to use the SystemV ABI causes a serious WTF reaction on my side.

tresf commented:
Here's [`libffi@56f7df7`](https://github.com/libffi/libffi/commit/56f7df711f70414d4f3663b34e54b122b38bab88): https://gist.githubusercontent.com/tresf/59f027cbff57d8537de0ac962257544f/raw/6b3d08bd59ec4f878e367083210c42fb0669235d/libffi@56f7df7.log

tresf commented:
Assuming there's no objections to the above new patches, I'm squashing them back into the PR.   Quoting:

> So the new fix is a bit more elegant.  Double conditional wasn't needed and if we wipe out `DLLCB` var, the `EXTRAOBJS+=` lower in the Makefile won't add it, so no stub `.o` is needed.  The custom precompiler flag is still needed (well, we could detect an aarch64 compiler, but it's more controlled this way and allows a farily quick toggle-off when mingw adds aarch64 support).
>
> After the squash, it's self-contained and easier to read as well.

tresf commented:
> I'd like to see the problem with libffi master resolved before this is merged.

This will also allow me to remove the `msvcc.sh` customizations from this PR.  Unfortunately, I'm not very familiar with the native components which newly introduced the bugs, but I'd be happy to run any tests.

I'll also shamelessly plug @dyorgio for assistance as we've worked together in the past and seems to be active here. 👋

matthiasblaesing commented:
From a first look of the build errors, something is fishy in the direct invoke code path. The "normal" JNA path goes through an invocation handler, while the direct approach registers at the JNI level. The error message:

```
    [junit] Testcase: testStringToIntArgumentConversion(com.sun.jna.DirectTypeMapperTest):      Caused an ERROR
    [junit] Native method linkage failed
    [junit] java.lang.Error: Native method linkage failed
    [junit]     at com.sun.jna.Native.registerMethod(Native Method)
    [junit]     at com.sun.jna.Native.register(Native.java:1854)
    [junit]     at com.sun.jna.Native.register(Native.java:1453)
    [junit]     at com.sun.jna.DirectTypeMapperTest$DirectTestLibraryString.<clinit>(DirectTypeMapperTest.java:69)
    [junit]     at com.sun.jna.DirectTypeMapperTest.testStringToIntArgumentConversion(DirectTypeMapperTest.java:144)
    [junit]     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    [junit]     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
    [junit]     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
```

is rooted here:

https://github.com/java-native-access/jna/blob/e6ab3ecdfa5482b41230fc28888df22da4d15bd4/native/dispatch.c#L3468-L3473

I would modify that code, to report, the requested ABI and the ffi_status (that might even be a good improvement to get into the codebase). That way we get some more information what is going on there.

tresf commented:
> I would modify that code, to report, the requested ABI and the `ffi_status` (that might even be a good improvement to get into the codebase). That way we get some more information what is going on there.

Happy to, any pointers?

matthiasblaesing commented:
Sure:

```diff
--- a/native/dispatch.c
+++ b/native/dispatch.c
@@ -3468,7 +3468,9 @@
   closure = ffi_closure_alloc(sizeof(ffi_closure), &code);
   status = ffi_prep_closure_loc(closure, closure_cif, dispatch_direct, data, code);
   if (status != FFI_OK) {
-    throwByName(env, EError, "Native method linkage failed");
+    char buf[MSG_SIZE];
+    const size_t count = snprintf(buf, MSG_SIZE, "Native method linkage failed (ffi_status: %d, abi: %d)", status, abi);
+    throwByName(env, EError, buf);
     goto cleanup;
   }
 ```

tresf commented:
Thanks, in each case:

```
java.lang.Error: Native method linkage failed (ffi_status: 2, abi: 2)
```

tresf commented:
> I'd like to see the problem with libffi master resolved before this is merged.

This will be rebased upon `libffi/libffi`'s `master` branch once https://github.com/libffi/libffi/pull/606 is merged upstream.

matthiasblaesing commented:
Ok - I think the wait is long enough - maybe it is time to consider just cherry picking the fix into our tree and be done with it. @tresf  what do you think?

tresf commented:
Since the last commits the following ARM64 changes have landed publicly... 
* Microsoft has released an [Insider Preview of Windows for ARM64](https://www.microsoft.com/en-us/software-download/windowsinsiderpreviewARM64) (registration required) finally allowing Windows on ARM64 to be emulated.
* Microsoft has released [x86_64 emulation for Windows ARM64](https://blogs.windows.com/windows-insider/2020/12/10/introducing-x64-emulation-in-preview-for-windows-10-on-arm-pcs-to-the-windows-insider-program/) in the Windows Developer Channel allowing us to test out the 64-bit toolchain (cygwin64 + Visual Studio x64).
* Parallels (for macOS) has released an [Apple M1 technical preview with Windows ARM64 support](https://my.parallels.com/desktop/beta) (registration required).

This gives me a unique opportunity to rebuild my system  from scratch to match the Windows build tutorial (64-bit toolchain) but this introduces a minor, unforeseen issue with cygwin documented here: 2714a01.  I've reported this issue to the cygwin developers.

Sorry for the TL;DR, but at time of writing this, I'm still encountering build failures with this setup.  When they're resolved, I'll get back to bumping libffi with the aforementioned (unmerged) patch and move forward.

tresf commented:
As it turns out, `gcc-core` is needed building "natively" on ARM64, as it toggles shared library support as well as offers other tools.  This is likely a bug with the build system, but for sake of timeliness (and lack of knowledge), I'm [amending the ARM64 table](https://github.com/java-native-access/jna/blob/5fbebe0e2034ec745aa1cedf82a4b4f99f2c4411/www/WindowsDevelopmentEnvironment.md#prerequisites) to include the appropriate mingw environment.

The list of tools which are reported as missing: `objdump`, `dlltool`, `ar`, `"@FILE"` (feature), `strip`, `ranlib`, `"parse dumpbin headers from msvcc.sh:` (feature), `"shared library support"` (feature).

The symptom is unobvious at first, the compile works as expected, but the `libjna.lib` has the wrong name and compilation fails (it's called `jna.lib` instead, "lib" prefix is missing).  I believe this is due to the "shared library support" missing from the compiler.

Anyway, I wanted to document this somewhere, as other projects relying on libffi and `msvcc.sh` may very well run into this down the road until the system is ported over to something that handles MSVC natively.

tresf commented:
Squashed and fast-forwarded, ready for final review.

The multiple force-pushes were needed to ensure a clean clone can still build binaries between changes, and then a final force push to bring the resulting `.jar` back in.

As an aside, I grew tired of the `.gitignore` rules breaking the `native/libffi` build system again and again, so I wrote a small batch script to bump `libffi` automatically, [`bump_ffi.bat`](https://gist.github.com/tresf/d36b4aedf0be45f3116575da0329731a) which fixes the `.gitignore` issues and automates the commit message.

matthiasblaesing commented:
I rebuild all native libraries with the new version, only skipping mac OS as that will be tackled with the alreay open PR, I also followed the notes about the windows build instructions and managed to rebuild all three windows native libraries. While doing this I slightly updated the documentation with my oberservation of the procedure.

Please have a look at this: https://github.com/matthiasblaesing/jna/tree/pr-1264. If you agree, I plan to merge that.

The native libraries can be found in the `lib/native` folder in the referenced commit. As an alternative here is a prebuild binary: 

[jna.zip](https://github.com/java-native-access/jna/files/5826814/jna.zip)

I plan to run the jna-platform tests with the updated base library, but I have yet to setup the necessary software.

tresf commented:
Cosmetic changes with the documentation seem sane, but the path changes don't work on my machines.  Commented on the commit in case you'd like to continue discussion there.

matthiasblaesing commented:
@tsref as discussed on the commit itself, I updated the branch with an updated suggestion for the wording of the windows instructions. The TL;DR version: I left you paths in place, but added a note that paths need to be updated depending on the exact software installed (I also removed the exact Java version, as long as JDK 8 is used, we are save).

I ran the platform tests on windows for 32 and 64bit. These came back clean, which is not surprising given, that appveyor is happy, in addition to the tests covered by appveyor on my machine (any machine that has Office installed) COM checks are run.

Please ensure, that the resulting binary can be run on aarch64, as I could only verify, that `file` reported the library as a PE file with arch aarch64. The binary can be found attached to the previous comment. (BTW: Is there an emulator available for Windows Aarch64?).

If you agree, I'll merge once Appveyor is happy.

tresf commented:
>  (BTW: Is there an emulator available for Windows Aarch64?).

QEMU can run the [full blown Windows ARM64 OS, but I've read it's tremendously slow](https://www.withinrafael.com/2018/02/12/boot-arm64-builds-of-windows-10-in-qemu/).  I also believe Microsoft offers a Windows 10X emulator, but it's been a very long time since I've fired it up and I'm unsure as to if the underlying OS is ARM64 or not.  I'll see if I have one around still.  It wasn't a very good OS for Desktop tasks last I tried (uses a weird Desktop sandbox) and the OS was too buggy for daily use, but that was about a year ago.

From what I can find most other emulators offered by Microsoft were old ARM32 layers, mostly for the phones.  I'll see what I can dig up.

>  in addition to the tests covered by appveyor on my machine (any machine that has Office installed) COM checks are run.

Would you like me to try the same on ARM64?

matthiasblaesing commented:
> > (BTW: Is there an emulator available for Windows Aarch64?).
> 
>  [QEMU Option]

Hehe, this reminds me about the pain when linux-mips64el and linux-ppc64el were build with full system emulation in Qemu....  Not fun :-)

> From what I can find most other emulators offered by Microsoft were old ARM32 layers, mostly for the phones. I'll see what I can dig up.
> 
> > in addition to the tests covered by appveyor on my machine (any machine that has Office installed) COM checks are run.
> 
> Would you like me to try the same on ARM64?

I have some fear here. My gut feeling is, that we could hit some hard assumptions, that don't hold anymore. Anyway - if can, it would be interesting, the final report would be great to see.

tresf commented:
In regards to Microsoft Office ARM64 integration... Office seems to use predominantly x86 binaries still, so that didn't add any more unit tests the ARM64 tests.  `file` reports `PE32 executable (GUI) Intel 80386, for MS Windows` (should be `PE32+ executable (DLL) (GUI) Aarch64, for MS Windows`).  This is using the latest Office 365 from Microsoft's website.

In regards to ARM64 emulation.... Windows 10X was a bust.  I'm unsure of the underlying architecture it's using, but what was very clear is that opening a command prompt on this OS isn't possible.  The Desktop emulation wouldn't even start the 7zip installer.

Microsoft claims Hyper-V will do architecture emulation when needed and there's evidence of this, such as the work from Xamarin team to port Android with hardware acceleration over to Hyper-V, but at time of writing this, the `.vhdx` image from MIcrosoft isn't bootable on Intel, leaving QEMU as the most documented/viable option.

matthiasblaesing commented:
I'll merged this in now - appveyor is happy, local tests with linux-x64, windows x86 and windows x64 came back clean. Thank you!

tresf commented:
Just an FYI on fast-forwarding/upstream status...
* :white_check_mark: https://github.com/libffi/libffi/pull/606 has now been merged into libffi, the cherry-pick (2732f2c) in this PR is no longer a one-off.
